### DISCUSSION TITLE
Clipboard content leaks to remote computer

### DISCUSSION THREAD

**Discussion by lucentwolf** [rustdesk/rustdesk#7346] [General]
### Bug Description

Irritating & posing leak-risks:

- A connects to B
- A performs a clipboard-copy (ctrl-c) outside Rustdesk (e.g. in editor)
- B performs a clipboard-paste (ctrl-v) and receives A's clipboard content.

Possibly linked to https://github.com/rustdesk/rustdesk/issues/7321

I consider this a *very* high risk since many password managers use clipboards & I don't like Rustdesk to leak it to computers/people I connect to

### How to Reproduce

see bug description

### Expected Behavior

A clipboard paste should only be using the local computer's clipboard content; that is

- A connects to B
- A performs ctrl-c outside Rustdesk
- If A performs ctrl-v in Rustdesk; A's clipboard content is used (on B's computer via Rustdesk)
- If B performs ctrl-v, B's clipboard content is used (on B's computer)

### Operating system(s) on local side and remote side

both Ubuntu 23.10 wayland 

### RustDesk Version(s) on local side and remote side

1.2.3

### Screenshots

-

**Comment by rustdesk** [comment]
All clipboard sync do in this way, including RDP, virtual machine etc.

**Reply by lucentwolf** [reply]
Absolutely not, sry. Just tested w/ Remmina/RDP:

- Remmina/RDP will only use my clipboard on the remote PC if I paste; a remote user paste will use the remote clipboard
- Rustdesk will use MY clipboard also if the REMOTE user pastes

**Reply by attikov** [reply]
Actually, it always uses the local clipboard; it's just automatically overwritten. However, we understand what you mean.

**Comment by lucentwolf** [comment]
...just noticed; Same is true in the other direction: A copy performed by the remote user will immediately populate my clipboard on my local computer!

Another Test even more worrying:

- A connects to B and C via Rustdesk
- User B makes a copy
- User C will have B's clipboard

Let me stress this: It's a security risk. Let's say in above scenario B and C are totally unrelated (e.g. two customers that I am connected to with two Rustdesk sessions); now B uses his pw-manager and C will have B's password in HIS clipboard... :-/

I might assume, that "syncing" would also spread the clipboard along connection strings; like if in above scenario B would connect to D; and D to E -> sry but this IS a security nightmare.

**Reply by xlionjuan** [reply]
I have encounter same behaviour, it is a little annoying.

**Reply by xlionjuan** [reply]
Anybody used KDE Plasma desktop, will know they have clipboard history, so I can see lot's unexpected contents in it.

**Comment by lucentwolf** [comment]
btw X or wayland makes no difference / can anyone test this behaviour on Windows?

**Reply by deep-soft** [reply]
I did noticed this behavior on Windows also.
I disabled clipboard for every remote connection, my clipboard leaked to all active connections.
If I need to copy something to remote clipboard: enable clipboard, copy to clipboard, paste from clipoard, disable clipboard... but I don't know other way for now.

**Comment by attikov** [comment]
My wish would be an option to completely disable clipboard auto-sync, along with two, easily accessible buttons for explicitly transferring the clipboard contents up or down. This should include metafile clipboards as well, such as for Excel.
Image transfers should remain disabled, just as they are currently.
Or make it selectable, specifying which types are allowed and which ones require confirmation, etc. I can provide ideas all day ;)

thx

**Comment by xlionjuan** [comment]
Another recommendation about clipboard sync

# Sync clipboard after connected
When I connected a device, this device's existing clipboard will override my clipboard, I think this clipboard monitor, should started after connected, not override at the begining.

**Comment by lucentwolf** [comment]
I think that the server (hosting a guest) will have to handle & decide when to sync - it must be aware of WHO does a copy / paste on it's desktop.

So when A (guest) connects to B's Desktop (host):

- Case 1: A does **copy** and/or **paste** outside Rustdesk -> 'local copy/paste' -> No SYNC
- Case 2: A does **copy** via Rustdesk on B's desktop -> B's Rustdesk will hand over the clipboard to A (only latest copy; no history); who then could use it locally or via Rustdesk on B's desktop.
- Case 3: A does a **paste** via Rustdesk on B's desktop -> B's Rustdesk will ask A's Rustdesk to hand over latest clipboard content; which could be a copy performed on B's desktop as well as a local copy - whatever happend most recently.
- Case 3: B does a **copy** and/or **paste** -> 'local copy/paste' -> no SYNC

This way only copy / paste that A performs on B's desktop will cross the connection; no unintentional content ever populated 'unrequested'... So IMHO a 'switch off clipboard' is an additional nice-to-have, not a core security feature. 

Now, only remaining question is whether or not it would make sense to allow B's copy (one that is done only AFTER receiving A as guest) to be handed over to A; so A could also use it (remotely or locally). 

Sidenote

Looking at Guacamole: When connected (no matter which protocol, RDP, SSH, whatever) there's an input box that a guests can open to read & populate clipboard content of the host's desktop. That's not very convenient (yet SAFE ^^); and probably due to using/dealing with how the BROWSER handles clipboard...

[EDIT: Typo]
